### PR TITLE
Notifications for friend requests

### LINKED ISSUE #250: Add support for notifications for friend requests and trips

### PR DESCRIPTION
**Pull request by BogdanCDev**
## Summary

This PR introduces enhanced notification behavior for friend requests in the `UserViewModel`.  
- When a user receives a new friend request, a notification is shown.  
- When one of the user's previously sent friend requests is accepted by another user, a separate notification is displayed to the sender and the accepted request is then removed.

This is a new feature addition, improving user awareness and interaction with friend requests.

## Changes

- **Models:**
  - Updated `FriendRequest` model to include an `accepted` field (`Boolean`) to track the acceptance state of a friend request.

- **Logic/Controllers:**
  - **`UserViewModel`:**
    - Retains the existing notification logic for new incoming friend requests.
    - Adds logic to detect when a previously sent friend request transitions to `accepted = true`.
    - Shows a notification when a sent request has been accepted, and then deletes the accepted request from Firestore.

- **Repositories:**
  - No structural changes to repositories, but `FriendRequestRepository` is now used to update and then delete the `accepted` friend requests.

- **Bug Fixes/Improvements:**
  - Ensures that after notifying the sender of an accepted friend request, the request is removed, preventing duplication or stale data.

## Checklist

Ensure all / most of these are checked before the pull request is submitted.
- [X] This PR resolves #250 
- [X] Tests have been added for new functionality.
- [X] Screenshots or UI changes have been attached 
- [X] Documentation has been updated

##  Testing

- **Manual Testing:**
    - [X] Tested on emulator / physical device.
- **Unit Tests**
    - [X] Added units tests for this

##  Related Issues/Tickets

Closes #250 

## Screenshots

https://github.com/user-attachments/assets/251d74d0-4f8e-454b-9837-a08f3ddee711

### CODE REVIEW

File: app/src/main/java/com/android/voyageur/App.kt [hunk 1]
```diff
@@ -34,7 +35,12 @@ fun VoyageurApp(placesClient: PlacesClient) {
   val navController = rememberNavController()
   val navigationActions = NavigationActions(navController)
   val tripsViewModel: TripsViewModel = viewModel(factory = TripsViewModel.Factory)
-  val userViewModel: UserViewModel = viewModel(factory = UserViewModel.Factory)
+  // Get the current context from the Compose hierarchy
+  val context = LocalContext.current
+
+  // Use UserViewModel.provideFactory(context) instead of UserViewModel.Factory
+  val userViewModel: UserViewModel = viewModel(factory = UserViewModel.provideFactory(context))
```
Comment (g0g05arui) [inline app/src/main/java/com/android/voyageur/App.kt:42]: why do we need the context in the userviewmodel? The view model should be agnostic of context which is related to the UI / composables.

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 2]
```diff
@@ -403,11 +417,34 @@ open class UserViewModel(
...
+                        // Remove the old request from this device's local state immediately
+                        _friendRequests.value =
+                            _friendRequests.value.filterNot { it.id == friendRequest.id }
+
+                        // After a short delay, delete the accepted request from Firestore
+                        // This gives the sender time to see the acceptance state and show a
+                        // notification
+                        viewModelScope.launch {
+                          kotlinx.coroutines.delay(1000)
+                          friendRequestRepository.deleteRequest(
```
Comment (g0g05arui) [inline app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:434]: Does this work if the other user is offline(without connection) / doesn't have the app open?

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 3]
```diff
@@ -579,39 +602,123 @@ open class UserViewModel(
...
+                // Check if there's a new request (list got bigger)
+                if (requests.size > oldRequests.size) {
+                  val oldRequestIds = oldRequests.map { it.id }.toSet()
+                  val newRequests = requests.filterNot { oldRequestIds.contains(it.id) }
+
+                  if (newRequests.isNotEmpty()) {
+                    val newRequest = newRequests.first()
+                    val senderUser = users.find { it.id == newRequest.from }
+                    val senderName = senderUser?.name ?: context?.getString(R.string.unknown)
+
```
Comment (g0g05arui) [inline app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:643]: you can not use the context here and just use the `stringResource` file which doesnt require a context

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 4]
```diff
@@ -579,39 +602,123 @@ open class UserViewModel(
...
+              val newlyAccepted =
+                  requests.filter { newReq ->
+                    newReq.accepted && oldRequestsMap[newReq.id]?.accepted == false
+                  }
+
+              if (newlyAccepted.isNotEmpty()) {
+                newlyAccepted.forEach { acceptedReq ->
+                  // The "to" user is the one who accepted
+                  getUsersByIds(listOf(acceptedReq.to)) { users ->
+                    val acceptorUser = users.firstOrNull()
```
Comment (g0g05arui) [inline app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:690]: Here you should use the `getUserById` with `acceptedReq.to` as the query by index (as the ID is used as the direct key) is way faster than the `arrayContains` function used in the `getUsersByIds`

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 5]
```diff
@@ -579,39 +602,123 @@ open class UserViewModel(
...
+                    val acceptorName = acceptorUser?.name ?: context?.getString(R.string.unknown)
+
+                    // Show the notification that the request was accepted
+                    context?.let {
+                      NotificationHelper.showNotification(
+                          context = context,
+                          notificationId = 1002,
+                          title = it.getString(R.string.friend_request_accepted),
```
Comment (g0g05arui) [inline app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:698]: Here 1002 is used as a hardcoded value. Maybe create a constant and explain it's value. 1002 Just seems random

**Comment by g0g05arui** [comment]
Also I see that we have not reached 80% line coverage on this PR. It might be beneficial to reach this level as this feature is critial to our application, so we should ensure that it works as expected. Otherwise, great job adding this feature! Good use of defensive programming concepts!

File: app/src/main/java/com/android/voyageur/App.kt (hunk 1 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/App.kt]: You are completely right! I updated the code to include 2 compostables that can be called in order for the use model to be agnostic to the context. Thanks for catching that!

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 2 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:437]: For now, since when we close the app we are no longer logged in you do not receive the notification that somebody accepted your friend request. But the 'kotlinx.coroutines.delay(1000)' makes sure that the friend request logic is updated correctly and that the notification is received if the app is opened. To implement notifications for when the app is not open I would need to use a push notification service like Firebase Cloud Messaging, which can be done, but I believe it should be the topic of another PR.

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 3 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt]: I do not believe that is possible since stringResource is a Composable and as such it can be called just from a Composable, but I handled this by using an auxiliary composable called AndroidStringProvider.

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 4 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:690]: I so not believe that getUserById is implemented. Am I missing something?

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 2 above)
Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:437]: Hmm, maybe you are right, but I was running under the impression that when we close the app the `Firebase.auth.uid` is still preserved so we don't necessarily log in (this is why the offline mode still works when you restart the app). 

Another thing you might have missed / might need to introduce in the next PR is that if you click on the notification for a friend request, the app should open directly on your profile so you can accept the friend request.

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 4 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:690]: Thank for the suggestion, you are absolutely right.

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 3 above)
Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt]: Ok, then your implementation is correct. Thank you for this quick reply!

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 2 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:437]: While you are right about the fact that Firebase.auth.uid is preserved, for the os to be able to handle background notifications, we need to implement a service like Firebase Cloud Messaging. Regarding your second point, that's for the suggestion, I will make sure to implement it as soon as Firebase Cloud Messaging is implemented.

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 5 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt]: Actually it is quite random, any unique id can be set in this field. I added a comment to the code to explain this.

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 6]
```diff
@@ -471,17 +508,20 @@ open class UserViewModel(
...
-    val Factory: ViewModelProvider.Factory =
-        object : ViewModelProvider.Factory {
-          @Suppress("UNCHECKED_CAST")
-          override fun <T : ViewModel> create(modelClass: Class<T>): T {
-            return UserViewModel(
-                UserRepositoryFirebase.create(),
-                friendRequestRepository = FriendRequestRepositoryFirebase.create())
-                as T
-          }
+    fun provideFactory(context: Context): ViewModelProvider.Factory {
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt]: Done. Thanks for pointing this out!

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 5 above)
Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt]: From what I read online, the ID should still be different from notfication to notification, so using the current timestamp and creating a (pseudo) random ID out of it might be the stolution. Take [this](https://stackoverflow.com/questions/12978184/android-get-unique-id-of-notification) as an example, it is in java but should be easily converted to kotlin.

Comment (g0g05arui) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt]: As if you use the same id twice, it just updates the notification, it doesn't create a new one (which would be the desired behaviour)

Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt]: When I though about this I thought that we should only keep the latest notifications such that the user is not getting too many notifications, but your suggestion seems to be actually better, thank you, I will implement it as soon as possible.

File: app/src/main/java/com/android/voyageur/ui/notifications/AndroidNotificationProvider.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package com.android.voyageur.ui.notifications
+
+import android.content.Context
+import androidx.core.app.NotificationCompat
+import com.android.voyageur.R
+
+open class AndroidNotificationProvider(private val context: Context) : NotificationProvider {
```
Comment (eleonoraka) [inline app/src/main/java/com/android/voyageur/ui/notifications/AndroidNotificationProvider.kt:7]: Thank you for implementing the AndroidNotificationProvider! It would be great to add some documentation to make the code easier to understand.

File: app/src/main/java/com/android/voyageur/ui/notifications/AndroidStringProvider.kt [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.android.voyageur.ui.notifications
+
+import android.content.Context
+
+class AndroidStringProvider(private val context: Context) : StringProvider {
```
Comment (eleonoraka) [inline app/src/main/java/com/android/voyageur/ui/notifications/AndroidStringProvider.kt:5]: I would suggest to add documentation here as well

File: app/src/androidTest/java/com/android/voyageur/ui/notifications/AndroidNotificationProvidertest.kt [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+  @After
+  fun tearDown() {
+    notificationManager.cancelAll()
+    activityScenario.close()
+  }
+
+  @Test
+  fun testNewFriendRequestNotification() {
+    // Wait for activity to be fully created
+    SystemClock.sleep(1000)
```
Comment (eleonoraka) [inline app/src/androidTest/java/com/android/voyageur/ui/notifications/AndroidNotificationProvidertest.kt:77]: Since we're removing all hardcoded strings, I suggest to use the same approach with "magic numbers" to maintain consistency throughout the code

File: app/src/main/java/com/android/voyageur/ui/notifications/AndroidNotificationProvider.kt [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  override fun showNewFriendRequestNotification(senderName: String) {
```
Comment (eleonoraka) [inline app/src/main/java/com/android/voyageur/ui/notifications/AndroidNotificationProvider.kt:8]: I see that the code structures for showNewFriendRequestNotification and showFriendRequestAcceptedNotification are very similar - you could refactor the code by extracting the common part into a separate function! This would reduce code duplication and also make it easier to add new types of notifications in the future. For example, you could create a 
`private showNotification(titleResId: Int, messageResId: Int, userName: String)` 
method that handles the common notification creation logic :)

**Review by eleonoraka — COMMENTED** [review]
Thank you for implementing the notifications! You did a great job and it enriches the app under many point of views. 
The overall code is well structured. I highlighted small issues and small enhancements you could address to make the code even better. 
You also did a good job with the tests. However I agree with @g0g05arui that we could improve the line coverage. To have better coverage I would suggest adding the following test cases:
- Edge cases in notification handling (null names, empty strings)
- Notification content and formatting
- Notification channel creation and configuration (if possible, I know there are some issues with testing some data structures and cases)
- Error scenarios in notification delivery (if possible, same as above)

Keep up the great work!

File: app/src/androidTest/java/com/android/voyageur/ui/notifications/AndroidNotificationProvidertest.kt (hunk 9 above)
Comment (BogdanCDev) [reply app/src/androidTest/java/com/android/voyageur/ui/notifications/AndroidNotificationProvidertest.kt:77]: They are not really the same thing since we remove hardcoded strings such that we can have support for multiple languages, but the time intervals in the code are needed in any situation and so I believe they can be hardcoded since they do not impact any UI elements. Also the code reference you put is from tests and we can implement them anyway we want.

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 5 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt]: Done!

File: app/src/main/java/com/android/voyageur/ui/notifications/AndroidStringProvider.kt (hunk 8 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/notifications/AndroidStringProvider.kt:10]: Done

File: app/src/main/java/com/android/voyageur/ui/notifications/AndroidNotificationProvider.kt (hunk 10 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/notifications/AndroidNotificationProvider.kt]: I am not sure I understand you comment since this is the current implementation.

**Comment by BogdanCDev** [comment]
> Thank you for implementing the notifications! You did a great job and it enriches the app under many point of views. The overall code is well structured. I highlighted small issues and small enhancements you could address to make the code even better. You also did a good job with the tests. However I agree with @g0g05arui that we could improve the line coverage. To have better coverage I would suggest adding the following test cases:
> 
> * Edge cases in notification handling (null names, empty strings)
> * Notification content and formatting
> * Notification channel creation and configuration (if possible, I know there are some issues with testing some data structures and cases)
> * Error scenarios in notification delivery (if possible, same as above)
> 
> Keep up the great work!

The notification handlers already have 100% line coverage. The since the target of 80% line coverage includes parts that will most likely be modified in a future pr, I believe that for now we can leave it like this for now and add more tests in that pr. Also the line coverage for user view model is already really high.

**Comment by BogdanCDev** [comment]
> Also I see that we have not reached 80% line coverage on this PR. It might be beneficial to reach this level as this feature is critial to our application, so we should ensure that it works as expected. Otherwise, great job adding this feature! Good use of defensive programming concepts!

Since, the notification handlers already have 100% line coverage and the target of 80% line coverage includes parts that will most likely be modified in a future pr, I believe that we can leave it like this for now and add more tests in that pr. Also the line coverage for user view model is already really high.

**Review by g0g05arui — APPROVED** [review]
